src/TuxState.js

```diff
@@ -0,0 +1,650 @@
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  unshiftState: function(propsToUnshift, callback) {
+    //throw error if propsToUnshift is undefined or not an object
+    invariantArgCheck(propsToUnshift);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToUnshift, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that current state at this key is an array
+      invariantArrayCheck(currentStateAtKey);
+      //check that there is only one value being pushed in
+      invariantValueCheck(newPropsAtKey);
+      //unshift new value into the corresponding array in the current state
+      currentStateAtKey.unshift(newPropsAtKey);
+      newState[key] = currentStateAtKey;
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //shiftState FUNCTION: match the deepest keys in the passed in object with the corresponding deepest keys in the current state and shift an item off the array in the corresponding key. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToShift OBJECT: required object argument with deepest keys being a boolean
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  shiftState: function(propsToShift, callback) {
+    //throw error if propsToShift is undefined or not an object
+    invariantArgCheck(propsToShift);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToShift, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that current state at this key is an array
+      invariantArrayCheck(currentStateAtKey);
+      //shift first value off the corresponding array in the current state
+      currentStateAtKey.shift();
+      newState[key] = currentStateAtKey;
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //spliceState FUNCTION: splice the values of the deepest keys in the current state with the array of arguments in the corresponding keys in the passed in object. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToSplice OBJECT: required object argument with deepest keys being an array of arguments to invoke splice with
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  spliceState: function(propsToSplice, callback) {
```

Comment:
Splice should probably also take some inputs defining the index to splice at and the number of array indices to remove when splicing.